pkg/services/hcloud/server/server.go

```diff
@@ -190,6 +217,198 @@ func (s *Service) handleBootStateUnset(ctx context.Context) (reconcile.Result, e
+
+	hm.SetBootState(infrav1.HCloudBootStateWaitForRescueRunningThenInstallImage)
+	return reconcile.Result{RequeueAfter: requeueAfterRebootToRescue}, nil
+}
+
+func (s *Service) handleBootStateWaitForRescueRunningThenInstallImage(ctx context.Context, _ *hcloud.Server) (reconcile.Result, error) {
+	hm := s.scope.HCloudMachine
+
+	if hm.Status.ActionIDRebootToRescue == 0 {
+		msg := "handleBootStateWaitForRescueRunningThenInstallImage ActionIdRebootToRescue not set? Can not continue. Provisioning Failed"
+		log.FromContext(ctx).Error(nil, msg)
+		s.scope.SetError(msg, capierrors.CreateMachineError)
+		return reconcile.Result{}, nil
+	}
+	if hm.Status.ActionIDRebootToRescue != actionDone {
+		action, err := s.scope.HCloudClient.GetAction(ctx, hm.Status.ActionIDRebootToRescue)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("GetAction failed: %w", err)
+		}
+
+		if action.Finished.IsZero() {
+			log.FromContext(ctx).Info("Waiting until Action RebootToRescue is finished", "action", action)
+			return reconcile.Result{RequeueAfter: requeueIntervalWaitForRescueRunning}, nil
+		}
+
+		err = action.Error()
+		if err != nil {
+			err = fmt.Errorf("action %+v failed: %w", action, err)
+			log.FromContext(ctx).Error(err, "")
+			s.scope.SetError(err.Error(), capierrors.CreateMachineError)
+			return reconcile.Result{}, nil
+		}
+
+		log.FromContext(ctx).Info("Action RebootToRescue is finished",
+			"actionDuration", action.Finished.Sub(action.Started),
+			"finishedSince", time.Since(action.Finished))
+
+		hm.Status.ActionIDRebootToRescue = actionDone
+	}
+
+	robotSecretName := s.scope.HetznerCluster.Spec.SSHKeys.RobotRescueSecretRef.Name
+	if robotSecretName == "" {
+		return reconcile.Result{}, fmt.Errorf("HetznerCluster.Spec.SSHKeys.RobotRescueSecretRef.Name is empty. Can not set hcloud machine into rescue-mode without ssh private key")
+	}
+
+	robotSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      robotSecretName,
+			Namespace: s.scope.Namespace(),
+		},
+	}
+
+	err := s.scope.Client.Get(ctx, client.ObjectKeyFromObject(robotSecret), robotSecret)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get secret %q: %w", robotSecretName, err)
+	}
+
+	if len(s.scope.HCloudMachine.Status.Addresses) == 0 {
+		return reconcile.Result{}, fmt.Errorf("HCloudMachine.Status.Addresses empty. Can not connect via ssh")
+	}
+
+	privateKey := string(robotSecret.Data[s.scope.HetznerCluster.Spec.SSHKeys.RobotRescueSecretRef.Key.PrivateKey])
+	if privateKey == "" {
+		return reconcile.Result{}, fmt.Errorf("key %q in secret %q is missing or empty. Failed to get ssh-private-key",
+			s.scope.HetznerCluster.Spec.SSHKeys.RobotRescueSecretRef.Key.PrivateKey,
+			robotSecretName)
+	}
+	ip := s.scope.HCloudMachine.Status.Addresses[0].Address
+
+	// Unfortunately the hcloud API does not provide the sshd hostkey of the rescue system.
+	// We need to trust the network. In theory a not man-in-the-middle attack is possible.
```

Comment:
"in theory a not ... " is this a correct sentence?

Reply:
fixed typo.